src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1120,6 +1135,78 @@ FinalizeRouterPlan(PlannedStmt *localPlan, CustomScan *customScan)
 }
 
 
+/*
+ * BlessRecordExpression ensures we can parse an anonymous composite type on the
+ * target list of a query that is sent to the worker.
+ *
+ * We cannot normally parse record types coming from the workers unless we
+ * "bless" the tuple descriptor, which adds a transient type to the type cache
+ * and assigns it a type mod value, which is the key in the type cache.
+ */
+static int32
+BlessRecordExpression(Expr *expr)
+{
+	int32 typeMod = -1;
+
+	if (IsA(expr, FuncExpr))
+	{
+		/*
+		 * Handle functions that return records on the target
+		 * list, e.g. SELECT function_call(1,2);
+		 */
+		Oid resultTypeId = InvalidOid;
+		TupleDesc resultTupleDesc = NULL;
+		TypeFuncClass typeClass;
+
+		/* get_expr_result_type blesses the tuple descriptor */
+		typeClass = get_expr_result_type((Node *) expr, &resultTypeId,
+										 &resultTupleDesc);
+		if (typeClass == TYPEFUNC_COMPOSITE)
+		{
+			typeMod = resultTupleDesc->tdtypmod;
+		}
+	}
+	else if (IsA(expr, RowExpr))
+	{
+		/*
+		 * Handle row expressions, e.g. SELECT (1,2);
+		 */
+		RowExpr *rowExpr = (RowExpr *) expr;
+		TupleDesc rowTupleDesc = NULL;
+		ListCell *argCell = NULL;
+		int currentResno = 1;
+
+		rowTupleDesc = CreateTemplateTupleDesc(list_length(rowExpr->args), false);
+
+		foreach(argCell, rowExpr->args)
+		{
+			Node *rowArg = (Node *) lfirst(argCell);
+			Oid rowArgTypeId = exprType(rowArg);
+			int rowArgTypeMod = exprTypmod(rowArg);
+
+			if (rowArgTypeId == RECORDOID)
+			{
+				/* ensure nested rows are blessed as well */
+				rowArgTypeMod = BlessRecordExpression((Expr *) rowArg);
+			}
+
+			TupleDescInitEntry(rowTupleDesc, currentResno, NULL,
+							   rowArgTypeId, rowArgTypeMod, 0);
+			TupleDescInitEntryCollation(rowTupleDesc, currentResno,
+										exprCollation(rowArg));
+
+			currentResno++;
+		}
+
+		BlessTupleDesc(rowTupleDesc);
```

Comment:
Note for other readers: I was worried about `blessing` too many records, but it turns out that Postgres is smart enough to create one entry for multiple calls, see `assign_record_type_typmod`